be/src/util/bthreads/future.h

```diff
@@ -0,0 +1,1105 @@
+namespace starrocks::bthreads {
+
+// Forward declarations.
+template <typename _Res>
+class future;
+
+template <typename _Res>
+class shared_future;
+
+template <typename _Signature>
+class packaged_task;
+
+template <typename _Res>
+class promise;
+
+/// Base class and enclosing scope.
+struct __future_base {
+    /// Base class for results.
+    struct _Result_base {
+        std::exception_ptr _M_error;
+
+        _Result_base(const _Result_base&) = delete;
+        _Result_base& operator=(const _Result_base&) = delete;
+
+        // _M_destroy() allows derived classes to control deallocation
+        virtual void _M_destroy() = 0;
+
+        struct _Deleter {
+            void operator()(_Result_base* __fr) const { __fr->_M_destroy(); }
+        };
+
+    protected:
+        _Result_base() noexcept = default;
+        virtual ~_Result_base() = default;
+    };
+
+    /// A unique_ptr for result objects.
+    template <typename _Res>
+    using _Ptr = std::unique_ptr<_Res, _Result_base::_Deleter>;
+
+    /// A result object that has storage for an object of type _Res.
+    template <typename _Res>
+    struct _Result : _Result_base {
+    private:
+        typename std::aligned_storage<sizeof(_Res), alignof(_Res)>::type _M_storage;
+        bool _M_initialized;
+
+    public:
+        typedef _Res result_type;
+
+        _Result() noexcept : _M_initialized() {}
+
+        ~_Result() override {
+            if (_M_initialized) _M_value().~_Res();
+        }
+
+        // Return lvalue, future will add const or rvalue-reference
+        _Res& _M_value() noexcept { return *static_cast<_Res*>(static_cast<void*>(&_M_storage)); }
+
+        void _M_set(const _Res& __res) {
+            ::new (static_cast<void*>(&_M_storage)) _Res(__res);
+            _M_initialized = true;
+        }
+
+        void _M_set(_Res&& __res) {
+            ::new (static_cast<void*>(&_M_storage)) _Res(std::move(__res));
+            _M_initialized = true;
+        }
+
+    private:
+        void _M_destroy() { delete this; }
+    };
+
+    // Keep it simple for std::allocator.
+    template <typename _Res, typename _Tp>
+    static _Ptr<_Result<_Res>> _S_allocate_result(const std::allocator<_Tp>& __a) {
+        return _Ptr<_Result<_Res>>(new _Result<_Res>);
+    }
+
+    // Base class for various types of shared state created by an
+    // asynchronous provider (such as a std::promise) and shared with one
+    // or more associated futures.
+    class _State_baseV2 {
+        typedef _Ptr<_Result_base> _Ptr_type;
+
+        enum _Status : unsigned { __not_ready, __ready };
+
+        _Ptr_type _M_result;
+        butil::atomic<int>* _M_status;
+        std::atomic_flag _M_retrieved{};
+        std::once_flag _M_once;
+
+    public:
+        _State_baseV2() noexcept : _M_result(), _M_status(nullptr), _M_once() {
+            _M_status = bthread::butex_create_checked<butil::atomic<int>>();
+            _M_status->store(_Status::__not_ready, butil::memory_order_release);
```

Comment:
Here is the mainly difference from the libstdc++: replaced the `__atomic_futex_unsigned` with `butex`